PG-1604: Improve last key LSN calculation logic

dutow opened the pull request:
Previosly we simply set the LSN for the new key to the first write
location.

This is however not correct, as there are many corner cases around this:
* recovery / replication might write old LSNs
* we can't handle multiple keys with the same TLI/LSN, which can happen
  with quick restarts without writes

To support this in this commit we modify the following:
* We only activate new keys outside crash recovery, or immediately if
  encryption is turned off
* We also take the already existing last key into account (if exists),
  and only activate a new key if we progressed past its start location

The remaining changes are just support infrastructure for this:

* Since we might rewrite old records, we use the already existing keys
  for those writes, not the active last keys
* We prefetch existing keys during initialization, so it doesn't
  accidentally happen in the critical section during a write

There is a remaining bug with stopping wal encryption, also mentioned in
a TODO message in the code. This will be addressed in a later PR as this
fix already took too long.



--

And two additional bugfixes, see the separate commits.

dutow committed:
Xlog encryption bugfix: offset calculation was off on TLI change

The min/max comparisons of LSNs assumed that everyting is in the same
timeline. In practice, with replication + recovery combinations, it is
possible that keys span at least 3 timelines, which means that this has
to be included in both combinations, as in other timelines, the
restrictions are less strict.

dAdAbird commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -340,7 +422,10 @@ tdeheap_xlog_seg_read(int fd, void *buf, size_t count, off_t offset,
-	TDEXLogCryptBuffer(buf, count, offset, tli, segno, segSize);
+	TDEXLogCryptBuffer(buf, buf, count, offset, tli, segno, segSize);
+
+	XLogSegNoOffsetToRecPtr(segno, offset + readsz, segSize, EncryptionState->LastReadOffset.lsn);
+	EncryptionState->LastReadOffset.tli = tli;
```
That should be inside `if (RecoveryInProgress())`
otherwise replication (walsender) would constantly update it

dutow replied:
Right. This is not noticable because we only update the key based on it once, but we don't have to update this variable later.

dAdAbird commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -234,7 +249,16 @@ TDEXLogSmgrInitWrite(bool encrypt_xlog)
+		// debug builds
+		// synchronization error? gcc bug? ub?
+		char		buf[33];
+		iv_prefix_debug(buf, EncryptionKey.key);
```
You're rewriting the EncryptionKey.key with the content of the buffer here. That's why maybe it "works", just because of the same key... I'm still trying to understand what's going on

dutow committed:
PG-1604: Improve last key LSN calculation logic

Previosly we simply set the LSN for the new key to the first write
location.

This is however not correct, as there are many corner cases around this:
* recovery / replication might write old LSNs
* we can't handle multiple keys with the same TLI/LSN, which can happen
  with quick restarts without writes

To support this in this commit we modify the following:
* We only activate new keys outside crash recovery, or immediately if
  encryption is turned off
* We also take the already existing last key into account (if exists),
  and only activate a new key if we progressed past its start location

The remaining changes are just support infrastructure for this:

* Since we might rewrite old records, we use the already existing keys
  for those writes, not the active last keys
* We prefetch existing keys during initialization, so it doesn't
  accidentally happen in the critical section during a write

There is a remaining bug with stopping wal encryption, also mentioned in
a TODO message in the code. This will be addressed in a later PR as this
fix already took too long.

dAdAbird commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +344,53 @@ static ssize_t
+		} else {
+			WALKeyCacheRec *last_key = pg_tde_get_last_wal_key();
+			if(last_key) {
+				if (wal_location_cmp(last_key->start, loc) < -8192) {
```
It's always going to be false, as wal_location_cmp() returns only -1, 0, or 1

AndersAstrand commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -415,8 +415,10 @@ TDEXLogCryptBuffer(void *buf, size_t count, off_t offset,
 				char		iv_prefix[16];
-				off_t		dec_off = XLogSegmentOffset(Max(data_start.lsn, curr_key->start.lsn), segSize);
-				off_t		dec_end = XLogSegmentOffset(Min(data_end.lsn, curr_key->end.lsn), segSize);
+				size_t		minlsn = data_end.tli < curr_key->end.tli ? data_end.lsn : Min(data_end.lsn, curr_key->end.lsn);
```
I think we should add a comment on why this is correct.

I might just be too dense to  understand the commit message. Which restrictions are less strict, and what are they less strict than?

AndersAstrand replied:
What I don't understand is when we want to use the larger of the two LSNs

dutow replied:
I added a comment and edited the variable names, is this better?

AndersAstrand replied:
Yes! Thank you. I just got a little tripped up by the caluculation for "end" being before the calculation for "start". It made a lot more sense to me when I realized that was the case :D. (my mind just read the first one and assume that would be the start).

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -415,16 +415,51 @@ TDEXLogCryptBuffer(void *buf, size_t count, off_t offset,
+				 * starts/ends in a TLI "within" the key, we can safely
+				 * decrypt/encrypt from the beginning / until the end, as it
+				 * is part of the key.
+				 */
```
I think what is important to explain is that a key can span into the next timeline but to lower LSNs, at least as long as they are part of the segment where the switch happened. So I think this comment should be focused on the important thing. More why and less what. :)

dutow replied:
Updated comment, please check.

jeltz replied:
Did you update it in the wrong commit?

jeltz reviewed (changes requested):
Review of the the bugifx comment.

dAdAbird commented on contrib/pg_tde/t/059_tde_2pc_replication.pl:
I think the test should be renamed as we abandoned number prefixes some time ago

dutow replied:
This is also a question for #528, I just copied the test from there.

dutow replied:
renamed, and also moved to the recovery directory similar to #528

dAdAbird commented on contrib/pg_tde/t/.pgtde.pm.swp:
Looks like a Vim leftover

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -226,6 +226,8 @@ void
 TDEXLogSmgrInitWrite(bool encrypt_xlog)
 {
 	WalEncryptionKey *key = pg_tde_read_last_wal_key();
+	WalLocation start = {.tli = 1,.lsn = 0};
```
This could be moved into the code block where it is used.

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -246,6 +248,14 @@ TDEXLogSmgrInitWrite(bool encrypt_xlog)
+	if (keys == NULL)
+	{
+		/* TODO cache is empty, try to preread keys from disk */
+		keys = pg_tde_fetch_wal_keys(start);
```
In a language with `_` which can be used for discarding an assiment this would have been perfect. Maybe just do not assign it to make it clear that we only do this to prime the cache. I suggest maybe:

```c
if (pg_tde_get_wal_cache_keys() == NULL)
{
    WalLocation start = {.tli = 1,.lsn = 0};

    pg_tde_fetch_wal_keys(start);
}
```

dutow replied:
this was a debugging leftover :)

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -349,20 +424,22 @@ tdeheap_xlog_seg_read(int fd, void *buf, size_t count, off_t offset,
-		keys = pg_tde_fetch_wal_keys(start);
+		pg_tde_fetch_wal_keys(start);
+
+		keys = pg_tde_get_wal_cache_keys();
```
If the cache is empty why can't we use the return value of `pg_tde_fetch_wal_keys()`? Maybe we should actually stop returning something from it and always do two fucntion calls?

dutow replied:
another debugging leftover, removed

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -425,11 +502,14 @@ TDEXLogCryptBuffer(void *buf, size_t count, off_t offset,
+				 * everything between two, including the entire TLI 4. For
+				 * example, TLI 4 LSN 1 and TLI 4 LSN 400 are both encrypted
+				 * with it, even through 1 is less than 100 and 400 is greater
+				 * than 200.
```
Shouldn't this be in the other commit?

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -465,17 +546,26 @@ TDEXLogCryptBuffer(void *buf, size_t count, off_t offset,
+				else
+				{
+					/* assert? */
+					dec_sz = 0;
```
Yes, we should assert. Just setting it to zero seems like it would hide bugs.

jeltz commented on src/bin/pg_basebackup/receivelog.c:
```diff
@@ -1131,7 +1134,8 @@ ProcessXLogDataMsg(PGconn *conn, StreamCtl *stream, char *copybuf, int len,
 #ifdef PERCONA_EXT
-		TDEXLogCryptBuffer(copybuf + hdr_len + bytes_written, bytes_to_write,
+		enc_buf = copybuf + hdr_len + bytes_written;
+		TDEXLogCryptBuffer(enc_buf, enc_buf, bytes_to_write,
```
Maybe use a code block to reduce the number of ifdefs?

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +336,64 @@ static ssize_t
-		pg_tde_wal_last_key_set_location(loc);
-		EncryptionKey.wal_start = loc;
-		TDEXLogSetEncKeyLocation(EncryptionKey.wal_start);
+	if (EncryptionKey.type != WAL_KEY_TYPE_INVALID && !lastKeyUsable)
```
Can't actually ever be invalid unless we have a bug? Maybe it should be an assert?

dutow replied:
done

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +336,64 @@ static ssize_t
+	{
+		WALKeyCacheRec *last_key = pg_tde_get_last_wal_key();
+
+		if (!crashRecovery || EncryptionKey.type == WAL_KEY_TYPE_UNENCRYPTED)
```
I would want a comment here explain why the unencrypted case does not need to care, because on top of my head I do not udnerstand why it would be special.

dutow replied:
It has to be special, we'll need more conditions around that - but that's not covered by the testcase, and I still have to figure out exactly what, that's why I left it as is and left a todo comment related to it.

jeltz replied:
Ok!

jeltz replied:
The more I think the less I get why it needs to be special. Don't we need to make sure old entries which are encryted stay encrypted with the same keys?

dAdAbird commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -246,6 +248,14 @@ TDEXLogSmgrInitWrite(bool encrypt_xlog)
+
+	if (keys == NULL)
+	{
+		/* TODO cache is empty, try to preread keys from disk */
```
I struggle to understand this comment

jeltz reviewed (changes requested):
I am still not a huge fan of rewinding and uisng the same key to write what we think should be new data but which technically does not have to be.

jeltz commented on contrib/pg_tde/xlogreader.c:
```diff
@@ -0,0 +1 @@
+../../src/backend/access/transam/xlogreader.c
```
Why the two symlinks?

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +336,66 @@ static ssize_t
+	pg_read_barrier();
+	last_key_loc.tli = TDEXLogGetEncKeyTli();
+
+	lastKeyUsable = (TDEXLogGetEncKeyLsn() != 0);
```
Shouldn't we only fetch this once? Not that it matters but it seems like a good practice.

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +336,66 @@ static ssize_t
-		XLogSegNoOffsetToRecPtr(segno, offset, segSize, loc.lsn);
+	last_key_loc.lsn = TDEXLogGetEncKeyLsn();
+	pg_read_barrier();
+	last_key_loc.tli = TDEXLogGetEncKeyTli();
```
Shouldn't this variable be called `write_key_loc`?

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +336,66 @@ static ssize_t
+	last_key_loc.tli = TDEXLogGetEncKeyTli();
+
+	lastKeyUsable = (TDEXLogGetEncKeyLsn() != 0);
+	afterLastKey = wal_location_cmp(last_key_loc, loc) <= 0;
```
Is it even necessary to check for this? Don't we write the WAL in a linear fashion so once recovery is done can we really write anything to old locations?

dutow replied:
yes, it is, because replicas might go back to earlier locations when rewriting an entire segment.

jeltz replied:
Even after recovery is done? Wow. I think this needs to be explained with a comment in the code.

dutow replied:
>Even after recovery is done?

Replicas are always in recovery, we can't rely on just checking recovery status. That was one of my earlier solutions, but replicas basically never generated new keys because of that (only in some corner cases)

So the current solution restricts the check for crash recovery, and generates a new key even on replicas after we progressed past the last key.

jeltz commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -299,26 +337,65 @@ static ssize_t
+	lastKeyUsable = (writeKeyLoc.lsn != 0);
+	afterWriteKey = wal_location_cmp(writeKeyLoc, loc) <= 0;
+
+
```
There is an extra empty line here.

dutow replied:
fixed
